### ISSUE TITLE
Add rule to detect duplicate reference links.

### ISSUE THREAD

**Issue by waylan** [markdownlint/markdownlint#137]
Sometimes the same reference key is defined multiple times in a document, however, the parser only uses the last definition. The linter should warn about this:

```
Some text with a [link][].

[link]: http://example.com

More text with a [link][].

[link]: http://otherexample.com
```

The fix would be to change one or both of the reference keys:

```
Some text with a [link][].

[link]: http://example.com

More text with a [link][other].

[other]: http://otherexample.com
```

Although less important, perhaps a warning would be appropriate when the same URL is assigned to multiple references: 

```
Some text with a [link][1].

[1]: http://example.com

More text with a [link][2].

[2]: http://example.com
```

The fix would be to use the same reference for both links:

```
Some text with a [link][1].

More text with a [link][1].

[1]: http://example.com
```

**Comment by d0ugal** [comment]
While the above examples are all valid, there is a slightly more dangerous case.

This example shows it, but pretend there are far more links and much more text so it isn't as easy to see. In the example, when I added the second link I accidentally changed both links to target the second URL.

``` markdown
Some text with a [link][].
More text with a [link][].

[link]: http://example.com
[link]: http://otherexample.com
```

**Comment by waylan** [comment]
> but pretend there are far more links and much more text so it isn't as easy to see.

@d0ugal makes a good point. In fact, my first example was demonstrating the same point and @d0ugal missed it (presumably) because I had a one-line paragraph between the two links with the same key. If you add multiple paragraphs, it becomes even more difficult.

Another possible rule would be to require all link definitions to be at the end of the document. Grouping them together makes it easier to see duplicates, but with a long list of definitions, duplicates are still easily missed. But, on a long document, scrolling between the link reference and the definition can be annoying, so maybe this rule should be off by default but available for those who want it.

A few other definition link rules I don't see are catching unused definitions (a definition exists but no link references it), or creating links with undefined references (a link references an undefined key).

In the end, I'm proposing five rules (which should apply to links and images):
1. No duplicate link definition keys
2. No duplicate link URLs
3. Unused link definition
4. Undefined reference
5. Group all link definitions at end

**Comment by dgw** [comment]
I'm surprised there hasn't been any activity on this in three years. Checking for unused link definitions was the very thing that prompted me to search for a "markdown linter" and find this project.

Despite the inactivity on this particular issue, it's good to see that this _project_ is still active, with a release made just last week. While I'm committing most of my FOSS time for the rest of this year to a major release of my main project (which I will not shill here 🙊), I'll add this issue to my list of things to look at after that. It would be a good excuse to dabble in Ruby.

**Comment by jaymzh** [comment]
The project was just recently resuscitated. :) I hit the low-hanging fruit that was blocking releases of various other software first. As I and the other new maintainers have time, we will try to start sifting through the backlog. Any help is welcome.

**Comment by dgw** [comment]
Sounds like where my project was last year. :) Reviving a dormant project is fun!

**Comment by guillaume-d** [comment]
The reference links should also not be broken, FYI see https://github.com/markdownlint/markdownlint/issues/95 for checking that (ATM only for inline-style links, but maybe reference-style links could be checked together with them).
